src/Lean/Server/InfoUtils.lean

```diff
@@ -189,4 +189,41 @@ where
       cs.any (hasNestedTactic pos tailPos)
     | _ => false
 
+/--
+Find info nodes that should be used for the term goal feature.
+
+The main complication concerns applications
+like `f a b` where `f` is an identifier.
+In this case, the term goal at `f`
+should be the goal for the full application `f a b`.
+
+Therefore we first gather the position of
+these head function symbols such as `f`,
+and later ignore identifiers at these positions.
```

Comment:
Oof, another case for making syntax tree instead of info tree the primary traversal. But I'm fine with the current implementation for now.